### PR TITLE
Media entity process plugin.

### PR DESCRIPTION
**Pull request by nayanalok**

### CODE REVIEW

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    return $html;
+  }
+
+  /**
+   * Get Media UUID by File ID.
+   */
+  protected function getMediaUuid(FileInterface $file, $media_bundle_field) {
+    $connection = \Drupal::database();
+    $query = $connection->select('media__' . $media_bundle_field, 'f');
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: `media__` ?

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * )
+ */
+class ApmMediaEntity extends ProcessPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($html, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+    // Values for the following variables are specified in the YAML file.
+    $images_source = $this->configuration['source_folder'];
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: what if all this config empty or not added in yaml? Add some exception handling.

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                  file_get_contents($filepath);
+              }
+              $new_destination = $destination . '/' . $row->getSourceProperty('id') . '-' . $filename;
+              if (!empty($file_contents)) {
+                if ($file = file_save_data($file_contents, $new_destination, FILE_EXISTS_REPLACE)) {
+                  // Create media entity using saved file.
+                  $media = Media::create([
+                    'bundle'      => $media_bundle,
+                    'uid'         => \Drupal::currentUser()->id(),
+                    'langcode'    => \Drupal::languageManager()->getDefaultLanguage()->getId(),
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: Inject user and langauage.

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                    'status'      => Media::PUBLISHED,
+                    $media_bundle_field => [
+                      'target_id' => $file->id(),
+                      'alt'       => !empty($tag_attributes[2][1]) ? Unicode::truncate(str_replace('"', '', $tag_attributes[2][1]), 512) : '',
+                      'title'     => !empty($tag_attributes[2][1]) ? Unicode::truncate(str_replace('"', '', $tag_attributes[2][1]), 1024) : '',
+                    ],
+                  ]);
+
+                  $media->save();
+                  $uuid = $this->getMediaUuid($file, $media_bundle_field);
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: what if no uuid is null?

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+              else {
+                $file_contents = (!empty($images_source)) ?
+                  file_get_contents($images_source . $filepath) :
+                  file_get_contents($filepath);
+              }
+              $new_destination = $destination . '/' . $row->getSourceProperty('id') . '-' . $filename;
+              if (!empty($file_contents)) {
+                if ($file = file_save_data($file_contents, $new_destination, FILE_EXISTS_REPLACE)) {
+                  // Create media entity using saved file.
+                  $media = Media::create([
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: create a helper method for media create so it not make transform method complex.

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                    $media_bundle_field => [
+                      'target_id' => $file->id(),
+                      'alt'       => !empty($tag_attributes[2][1]) ? Unicode::truncate(str_replace('"', '', $tag_attributes[2][1]), 512) : '',
+                      'title'     => !empty($tag_attributes[2][1]) ? Unicode::truncate(str_replace('"', '', $tag_attributes[2][1]), 1024) : '',
+                    ],
+                  ]);
+
+                  $media->save();
+                  $uuid = $this->getMediaUuid($file, $media_bundle_field);
+                  $replace_image = '<p><drupal-entity
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: create a helper for uuid creation as well.

File: src/Plugin/migrate/process/ApmMediaEntity.php (hunk 1 above)
Comment (nayanalok) [reply src/Plugin/migrate/process/ApmMediaEntity.php]: Its a table format for media bundle field.

File: src/Plugin/migrate/process/ApmMediaEntity.php (hunk 2 above)
Comment (nayanalok) [reply src/Plugin/migrate/process/ApmMediaEntity.php]: included validations

File: src/Plugin/migrate/process/ApmMediaEntity.php (hunk 5 above)
Comment (nayanalok) [reply src/Plugin/migrate/process/ApmMediaEntity.php]: created helper method createMediaEntity()

File: src/Plugin/migrate/process/ApmMediaEntity.php (hunk 4 above)
Comment (nayanalok) [reply src/Plugin/migrate/process/ApmMediaEntity.php]: included condition to check null

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+   * @var \Drupal\Core\Session\AccountProxyInterface
+   */
+  protected $currentUser;
+
+  /**
+   * User account service.
+   *
+   * @var \Drupal\Core\Language\LanguageManagerInterface
+   */
+  protected $defaultLanguage;
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: `$langugaeManager`

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+  /**
+   * ApmMediaEntity constructor to check for required fields.
+   * @param array $configuration
+   * @param $pluginId
+   * @param $pluginDefinition
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: add comment for above three params as well.

File: src/Plugin/migrate/process/ApmMediaEntity.php [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+   * @param ModuleHandlerInterface $moduleHandler
+   *   Module handler to check if module exist
+   * @param AccountProxyInterface $currentUser
+   *   Get current active user
+   * @param LanguageManagerInterface $defaultLanguage
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMediaEntity.php]: `$language_manager`

**Comment by joshirohit100** [comment]
PR looks fine to me except minor comments. 

cc @amitgoyal  for review and merge and then push to private repo.

**Pull request merged**